### PR TITLE
Space behavior QoL updates for Han shape-based layout

### PR DESCRIPTION
**Pull request by moonbeamcelery**
Separate the space behavior of the new nlp providers and the regular Latin ones, and expose setting nlp providers to the subtype editor. These allow Latin layouts and the new layout to coexist.

TODO: switching subtypes with different nlp providers currently does not reset suggestions, so you would still have suggestions from another language until you type or move your cursor.

### CODE REVIEW

**Comment by moonbeamcelery** [comment]
Note: also included is avoiding suggestions when the composed text is empty (instead of suggesting characters for a, aa, aaa, etc.)

**Review by Waelwindows — COMMENTED** [review]
This would better as a separate PR

File: app/src/main/kotlin/dev/patrickgold/florisboard/ime/editor/EditorInstance.kt [hunk 1]
```diff
@@ -502,17 +504,17 @@ class EditorInstance(context: Context) : AbstractEditorInstance(context) {
...
-        // return textBefore.isNotEmpty() &&
-        //     (punctuationRule.symbolsPrecedingPhantomSpace.contains(textBefore[textBefore.length - 1]) ||
-        //         textBefore[textBefore.length - 1].isLetterOrDigit()) &&
-        //     (punctuationRule.symbolsFollowingPhantomSpace.contains(text[0]) || text[0].isLetterOrDigit())
+         val content = activeContent
+         val selection = content.selection
+         if (!(isActive || forceActive) || selection.isNotValid || selection.start <= 0 || text.isEmpty()) return false
+         val textBefore = content.getTextBeforeCursor(1)
+         val punctuationRule = nlpManager.getActivePunctuationRule()
+        // TODO: maybe put this recurring condition in a function that determines CJK
```
Comment (Waelwindows) [inline app/src/main/kotlin/dev/patrickgold/florisboard/ime/editor/EditorInstance.kt]: It's better to add this function in this PR

**Review by Waelwindows — CHANGES_REQUESTED** [review]

File: app/src/main/kotlin/dev/patrickgold/florisboard/ime/editor/EditorInstance.kt (hunk 1 above)
Comment (Waelwindows) [reply app/src/main/kotlin/dev/patrickgold/florisboard/ime/editor/EditorInstance.kt]: Perhaps look into adding it to `FlorisLocale.kt`?

**Comment by moonbeamcelery** [comment]
> This would better as a separate PR

You mean the one where avoiding suggestions when composing text is empty?

**Comment by Waelwindows** [comment]
> > This would better as a separate PR
> 
> You mean the one where avoiding suggestions when composing text is empty?

Sorry for not being clear. I'm new to GitHub's review process. 

I meant the suggestion provider UI

File: app/src/main/kotlin/dev/patrickgold/florisboard/ime/keyboard/KeyboardManager.kt [hunk 2]
```diff
@@ -549,8 +550,11 @@ class KeyboardManager(context: Context) : InputKeyEventReceiver {
                 }
             }
         }
-        // TODO: Properly handle not adding a space when commiting suggestions in CJK subtypes
-        // editorInstance.commitText(KeyCode.SPACE.toChar().toString())
+        // TODO: maybe put this recurring condition in a function that determines CJK
```
Comment (moonbeamcelery) [inline app/src/main/kotlin/dev/patrickgold/florisboard/ime/keyboard/KeyboardManager.kt]: TODO: delet this

**Comment by moonbeamcelery** [comment]
Updated with suggested changes. The editor change will be another one, although I couldn't get it to work now that the db is dependent on primaryLocale.variant

**Comment by Waelwindows** [comment]
> Updated with suggested changes. The editor change will be another one, although I couldn't get it to work now that the db is dependent on primaryLocale.variant

Did you try with the zh-RC-zhengma layout?

**Comment by moonbeamcelery** [comment]
> > Updated with suggested changes. The editor change will be another one, although I couldn't get it to work now that the db is dependent on primaryLocale.variant
>
> Did you try with the zh-RC-zhengma layout?

~I couldn't find it~
Ok I found it. I left a comment in your pr regarding this.

Edit: confirmed it works. Thank you!

**Review by Waelwindows — APPROVED** [review]

**Comment by moonbeamcelery** [comment]
Hello again Waelwindows, is there anything I can help with getting your PR merged? Would making a Chinese layout be helpful for example? I really appreciate your work and would love to see the zhengma input feature on the main repo soon!

**Comment by Waelwindows** [comment]
Adding more layouts would be great! 

Wish I could work more on FlorisBoard but as uni student, exam season never seems to be away for too long :sweat_smile:.

For now, I think it's good to get other tables from [fcitx5-tables-extra](https://github.com/fcitx/fcitx5-table-extra) as the data from there is from the public domain (or so they claim).

I should probably do that for zhengma as well, since I got that from `ibus-zhengma` which i'm not sure uses public domain tables.

Perhaps the above as different PRs? Hopefully i'll respond to them on time

**Comment by Waelwindows** [comment]
I have looked into the PR and it seems that we've forgotten to invert the condition :joy:.

Although there's another problem now, after committing the suggestion, a space is added when you write a letter. Gotta look into that.

**Comment by Waelwindows** [comment]
Got it, kinda weird that I had to negate all uses of it, perhaps the meaning was flipped in my implementation?

**Comment by moonbeamcelery** [comment]
Oops, sorry I didn't see this one, lol.

Yeah, it used to be `isCJK` which would indeed be `!supportAutoSpace`.

**Comment by moonbeamcelery** [comment]
> Adding more layouts would be great!
> 
> Wish I could work more on FlorisBoard but as uni student, exam season never seems to be away for too long sweat_smile.
> 
> For now, I think it's good to get other tables from [fcitx5-tables-extra](https://github.com/fcitx/fcitx5-table-extra) as the data from there is from the public domain (or so they claim).
> 
> I should probably do that for zhengma as well, since I got that from `ibus-zhengma` which i'm not sure uses public domain tables.
> 
> Perhaps the above as different PRs? Hopefully i'll respond to them on time

Got it, I'll try some of these when I have time.

**Pull request merged**